api/datacontext/attrs/featuregatesattr/attr.go

```diff
@@ -0,0 +1,199 @@
+package featuregatesattr
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/mandelsoft/goutils/general"
+	"sigs.k8s.io/yaml"
+
+	"ocm.software/ocm/api/datacontext"
+	"ocm.software/ocm/api/utils/runtime"
+)
+
+const (
+	ATTR_KEY   = "ocm.software/ocm/feature-gates"
+	ATTR_SHORT = "featuregates"
+)
+
+func init() {
+	datacontext.RegisterAttributeType(ATTR_KEY, AttributeType{}, ATTR_SHORT)
+}
+
+type AttributeType struct{}
+
+func (a AttributeType) Name() string {
+	return ATTR_KEY
+}
+
+func (a AttributeType) Description() string {
+	return `
+*featuregates* Enable/Disable optional features of the OCM library. Optional,
+particular features modes and attributes can be configured, if supported by 
+the feature implementation.
+`
+}
+
+func (a AttributeType) Encode(v interface{}, marshaller runtime.Marshaler) ([]byte, error) {
+	switch t := v.(type) {
+	case *Attribute:
+		return json.Marshal(v)
+	case string:
+		_, err := a.Decode([]byte(t), runtime.DefaultYAMLEncoding)
+		if err != nil {
+			return nil, err
+		}
+		return []byte(t), nil
+	case []byte:
+		_, err := a.Decode(t, runtime.DefaultYAMLEncoding)
+		if err != nil {
+			return nil, err
+		}
+		return t, nil
+	default:
+		return nil, fmt.Errorf("feature gate config required")
+	}
+}
+
+func (a AttributeType) Decode(data []byte, unmarshaller runtime.Unmarshaler) (interface{}, error) {
+	var c Attribute
+	err := yaml.Unmarshal(data, &c)
+	if err != nil {
+		return nil, err
+	}
+	return &c, nil
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+const FEATURE_DISABLED = "off"
+
+type Attribute struct {
+	lock sync.Mutex
+
+	Features map[string]*FeatureGate `json:"features"`
+}
+
+// FeatureGate store settings for a particular feature gate.
+// To be extended by additional config possibility.
+// Default behavior is to be enabled if entry is given
+// for a feature name and mode is not equal *off*.
+type FeatureGate struct {
+	Mode       string                     `json:"mode"`
```

Comment:
Generally in favor of this, but why choose to have a Mode setting for the Gate if you already have attributes? Shouldn't the mere presence of the gate determine if its enabled or disabled?

Reply:
The idea of the mode is to have a fixed/non-optional config field usable to enable/disable a feature and control an implementation mode (if required).

The attributes are completely optional and intended to pass arbitrary information to the feature. (An on/off switch does not use attributes at all (omitempty))

Reply:
> The idea of the mode is to have a fixed/non-optional config field usable to enable/disable a feature and control an implementation mode (if required). The value off is always used to enforce disabling the feature (in case it is enabled by default).

Could you give me an example for an "implementation mode" that is not on/off? Ive never seen it anywhere in production and was assuming that even if there was a non on/off, it would be part of the attributes.